src/Dfc.CourseDirectory.WebV2/Features/Venues/AddVenue/AddVenueController.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System.Linq;
+using System.Threading.Tasks;
+using Flurl;
+using FormFlow;
+using MediatR;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Dfc.CourseDirectory.WebV2.Features.Venues.AddVenue
+{
+    [Route("venues/add")]
+    [JourneyMetadata(
+        journeyName: "AddVenue",
+        stateType: typeof(AddVenueJourneyModel),
+        appendUniqueKey: true,
+        requestDataKeys: "providerId?")]
+    [RequireProviderContext]
+    public class AddVenueController : Controller
+    {
+        private const string ReturnUrlJourneyInstancePropertyKey = "returnUrl";
+
+        private readonly IMediator _mediator;
+        private readonly JourneyInstanceProvider _journeyInstanceProvider;
+        private readonly IProviderContextProvider _providerContextProvider;
+        private JourneyInstance<AddVenueJourneyModel> _journeyInstance;
+
+        public AddVenueController(
+            IMediator mediator,
+            JourneyInstanceProvider journeyInstanceProvider,
+            IProviderContextProvider providerContextProvider)
+        {
+            _mediator = mediator;
+            _journeyInstanceProvider = journeyInstanceProvider;
+            _providerContextProvider = providerContextProvider;
+            _journeyInstance = _journeyInstanceProvider.GetInstance<AddVenueJourneyModel>();
+        }
+
+        [HttpGet("")]
+        public IActionResult Index([FromQuery] string postcode, [FromQuery] string returnUrl)
+        {
+            _journeyInstance = _journeyInstanceProvider.GetOrCreateInstance(
+                () => new AddVenueJourneyModel(),
+                new PropertiesBuilder()
+                    .Add(ReturnUrlJourneyInstancePropertyKey, returnUrl)
+                    .Build());
+
+            if (!_journeyInstanceProvider.IsCurrentInstance(_journeyInstance))
+            {
+                return RedirectToAction()
```

Comment:
Surely we don't want to redirect to itself here, we want to show a meaningful error, perhaps `BadRequest`?